### PR TITLE
Add option for clearing track's waveform

### PR DESCRIPTION
**Pull request by ninomp**
This PR fixes https://bugs.launchpad.net/mixxx/+bug/1014449

However, I haven't managed to get this fully functional: clearing waveform doesn't work when a small number of tracks (< 5) are selected, but does work when more tracks are selected. By looking at Mixxx log, it seems to me that this is a caching issue and a small number of changes is not (immediately) written to database. Can someone of the more experienced Mixxx developers please help me with this? Thanks in advance.

### CODE REVIEW

File: src/library/dao/analysisdao.cpp [hunk 1]
```diff
@@ -323,13 +323,21 @@ void AnalysisDao::saveTrackAnalyses(TrackInfoObject* pTrack) {
...
+        !pWaveSummary || pWaveSummary->getDataSize() == 0) {
+        bool success = deleteAnalysesForTrack(trackId);
+        qDebug() << (success ? "Successfully deleted" : "Failed to delete")
+                 << "waveform analysis for trackId" << trackId;
         return;
     }
 
-    TrackId trackId(pTrack->getId());
+    // Don't try to save non-dirty waveforms.
+    if (!pWaveform->isDirty() || !pWaveSummary->isDirty()) {
```
Comment (daschuer) [inline src/library/dao/analysisdao.cpp]: Why was that moved to here? 
Can it happen that we have to clear a non dirty waveform with DataSize = 0?
When will the data size drop to 0?

File: src/trackinfoobject.cpp [hunk 2]
```diff
@@ -722,20 +722,28 @@ QString TrackInfoObject::getURL() {
 }
 
 ConstWaveformPointer TrackInfoObject::getWaveform() {
+    QMutexLocker lock(&m_qMutex);
```
Comment (daschuer) [inline src/trackinfoobject.cpp]: Isn't returning a shared pointer should be already thread save?

File: src/widget/wtracktableview.cpp [hunk 3]
```diff
@@ -1597,6 +1603,22 @@ void WTrackTableView::slotClearBeats() {
...
+    TrackModel* trackModel = getTrackModel();
+    if (trackModel == NULL) {
+        return;
+    }
+
+    QModelIndexList selectedIndices = selectionModel()->selectedRows();
+    foreach (QModelIndex index, selectedIndices) {
+        TrackPointer pTrack = trackModel->getTrack(index);
+        if (pTrack) {
+            pTrack->setWaveform(WaveformPointer());
```
Comment (daschuer) [inline src/widget/wtracktableview.cpp]: This means, you dereference the Waveform, where the TrackInfo member variable is pointing to. 
This does not effect the references that are already spread though Mixxx via getWaveform(). 

So you need to inform all getWaveform() users, that the waveform hold by the original pointer is now invalid. 

Maybe we can set an atomic flag in the original waveform ..

File: src/trackinfoobject.cpp [hunk 4]
```diff
@@ -722,20 +722,28 @@ QString TrackInfoObject::getURL() {
...
     return m_waveform;
 }
 
 void TrackInfoObject::setWaveform(ConstWaveformPointer pWaveform) {
+    QMutexLocker lock(&m_qMutex);
     m_waveform = pWaveform;
+    setDirty(true);
+    lock.unlock();
     emit(waveformUpdated());
```
Comment (daschuer) [inline src/trackinfoobject.cpp]: Are all users of getWaveform listening to this signal?

**Comment by uklotzde** [comment]
The TrackDAO has an internal LRU cache of 5 tracks.

File: src/trackinfoobject.cpp (hunk 2 above)
Comment (ninomp) [inline src/trackinfoobject.cpp]: I'm not sure. Almost all the other geters and setters use locks, so I added one here also.

File: src/library/dao/analysisdao.cpp (hunk 1 above)
Comment (ninomp) [inline src/library/dao/analysisdao.cpp]: Well, the original idea was that a NULL pointer to waveform (or an empty waveform) mean that waveform should be cleared, but adding a flag to waveform seems to be a better/cleaner/more elegant solution.

File: src/trackinfoobject.cpp (hunk 4 above)
Comment (ninomp) [inline src/trackinfoobject.cpp]: By grepping thru code, I found that `AnalyzerWaveform` is responsible for loading of an existing waveform in case the track is missing waveform.

File: src/widget/wtracktableview.cpp (hunk 3 above)
Comment (ninomp) [inline src/widget/wtracktableview.cpp]: I've done some thinking (that's why I'm not answering for two days :smile:) and I like your idea for adding a flag to `Waveform`. IMHO it is much cleaner than having a NULL pointer. But how do I name it? `m_bNeedsToBeCleared` seems a little too long and `m_bInvalid` too ambiguous (`Waveform` already has an `isValid` method).

Comment (daschuer) [inline src/widget/wtracktableview.cpp]: m_bNeedsToBeCleared is not that bad ;-) If the underlying concept is proved.

**Comment by ninomp** [comment]
@daschuer, @uklotzde thank you for your help! :-) Although the discussed idea was to add a flag to Waveform, I decided to add a flag in TIO instead. Now, I would like to merge this and start working on a fix for https://bugs.launchpad.net/mixxx/+bug/1457746.

File: src/trackinfoobject.cpp (hunk 2 above)
Comment (uklotzde) [inline src/trackinfoobject.cpp]: Locking is not necessary here. Accessing the native pointer wrapped in QSharedPointer is thread safe. See also: Qt Docs > QSharedPointer > Thread-Safety

File: src/trackinfoobject.cpp [hunk 5]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
...
+    }
+
+    QMutexLocker lock(&m_qMutex);
     m_waveform = pWaveform;
+    markDirtyAndUnlock(&lock);
     emit(waveformUpdated());
 }
 
 ConstWaveformPointer TrackInfoObject::getWaveformSummary() const {
+    QMutexLocker lock(&m_qMutex);
```
Comment (uklotzde) [inline src/trackinfoobject.cpp]: Same here, no need for locking.

File: src/trackinfoobject.cpp [hunk 6]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
...
     return m_waveform;
 }
 
 void TrackInfoObject::setWaveform(ConstWaveformPointer pWaveform) {
+    if (m_waveform == pWaveform) {
+        return;
+    }
+
+    QMutexLocker lock(&m_qMutex);
     m_waveform = pWaveform;
```
Comment (uklotzde) [inline src/trackinfoobject.cpp]: Should we also fix the names of those members according to the naming conventions?

m_waveform(Summary) -> m_pWaveform(Summary)

The code looks a little odd with the parameter using the 'p' prefix, while the member does not.

File: src/trackinfoobject.cpp [hunk 7]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
...
     emit(waveformUpdated());
 }
 
 ConstWaveformPointer TrackInfoObject::getWaveformSummary() const {
+    QMutexLocker lock(&m_qMutex);
     return m_waveformSummary;
 }
 
 void TrackInfoObject::setWaveformSummary(ConstWaveformPointer pWaveform) {
+    if (m_waveformSummary == pWaveform) {
```
Comment (uklotzde) [inline src/trackinfoobject.cpp]: Here you rely on the thread-safety guarantees of QSharedPointer. Perfectly ok, no need for locking :)

But we should apply those rules consistently and remove the extra locks from the getters (see above)

File: src/trackinfoobject.cpp [hunk 8]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
...
+    }
+
+    QMutexLocker lock(&m_qMutex);
     m_waveformSummary = pWaveform;
+    markDirtyAndUnlock(&lock);
     emit(waveformSummaryUpdated());
 }
 
+bool TrackInfoObject::isClearWaveformRequested() const {
+    return m_bClearWaveformRequested;
```
Comment (uklotzde) [inline src/trackinfoobject.cpp]: Here a lock is required, even if it is only a primitive boolean value!

File: src/trackinfoobject.cpp [hunk 9]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
...
+}
+
+void TrackInfoObject::setClearWaveformRequested(bool requested) {
+    m_bClearWaveformRequested = requested;
```
Comment (uklotzde) [inline src/trackinfoobject.cpp]: Same here, locking is required

File: src/widget/wtracktableview.cpp [hunk 10]
```diff
@@ -1597,6 +1603,21 @@ void WTrackTableView::slotClearBeats() {
     }
 }
 
+void WTrackTableView::slotClearWaveform() {
+    TrackModel* trackModel = getTrackModel();
+    if (trackModel == NULL) {
```
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: C++11: Use nullptr instead of NULL

**Comment by ninomp** [comment]
Thank you @uklotzde for your review. I have addressed your review comments.

File: src/library/dao/analysisdao.cpp [hunk 11]
```diff
@@ -323,14 +323,26 @@ void AnalysisDao::saveTrackAnalyses(TrackInfoObject* pTrack) {
     ConstWaveformPointer pWaveform = pTrack->getWaveform();
     ConstWaveformPointer pWaveSummary = pTrack->getWaveformSummary();
 
+    TrackId trackId(pTrack->getId());
+
+    // Delete waveform analysis if track was requested to have its waveform cleared.
+    if (pTrack->isClearWaveformRequested()) {
```
Comment (uklotzde) [inline src/library/dao/analysisdao.cpp]: The code in both AnalyzerWaveform and AnalysisDao for clearing the waveform if requested looks very similar. I would recommend to encapsulate the code from AnalysisDao in a function:

bool AnalysisDao::clearWaveformIfRequested(TrackInfoObject* pTrack)

Then from AnalyzerWaveform::loadStored() you could simply invoke pAnalysisDao->clearWaveformIfRequested(tio) before obtaining the waveform pointers from the TIO.

File: src/library/dao/analysisdao.cpp [hunk 12]
```diff
@@ -323,14 +323,26 @@ void AnalysisDao::saveTrackAnalyses(TrackInfoObject* pTrack) {
...
+        bool success = deleteAnalysesForTrack(trackId);
+        qDebug() << (success ? "Successfully deleted" : "Failed to delete")
+                 << "waveform analysis for trackId" << trackId;
+        if (success) {
```
Comment (uklotzde) [inline src/library/dao/analysisdao.cpp]: If you don't reset the flag in case deleteAnalysesForTrack(trackId) fails you will repeatedly enter this code path without actually resetting the flag. Maybe the analysis has already been deleted? I can't think of any other reason why deletion should fail. You might log a warning with qWarning() on failure, but the TIO's flag should always be reset. The proposed function should return true if the flag has been reset and false otherwise.

File: src/library/dao/analysisdao.cpp (hunk 11 above)
Comment (ninomp) [inline src/library/dao/analysisdao.cpp]: Yes, you are right, but I'm planning to implement automatic clear of waveforms, beats and keys of track being loaded if Mixxx detects that audio of the track has changed, so I need a name that will cover both (automatic and manual) clearing. Do you mind if I do this then?

Comment (uklotzde) [inline src/library/dao/analysisdao.cpp]: Sure :) Thanks for your work, Nino! It's a pleasure to review your commits, 
well done.

Comment (ninomp) [inline src/library/dao/analysisdao.cpp]: Thank you! :smile: I'm glad you like it.

File: src/library/dao/analysisdao.cpp (hunk 12 above)
Comment (ninomp) [inline src/library/dao/analysisdao.cpp]: You are right! I will fix this immediatelly.

**Comment by daschuer** [comment]
Is the waveform still gone if you: 
Clear Waveform. 
Close Mixxx
Restart Mixxx

File: src/trackinfoobject.cpp [hunk 13]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
 }
 
 ConstWaveformPointer TrackInfoObject::getWaveform() {
-    return m_waveform;
+    return m_pWaveform;
 }
 
 void TrackInfoObject::setWaveform(ConstWaveformPointer pWaveform) {
-    m_waveform = pWaveform;
+    if (m_pWaveform == pWaveform) {
```
Comment (daschuer) [inline src/trackinfoobject.cpp]: The check has to be done inside the lock.

File: src/trackinfoobject.cpp [hunk 14]
```diff
@@ -738,23 +739,51 @@ QString TrackInfoObject::getURL() {
...
 }
 
 ConstWaveformPointer TrackInfoObject::getWaveformSummary() const {
-    return m_waveformSummary;
+    return m_pWaveformSummary;
 }
 
 void TrackInfoObject::setWaveformSummary(ConstWaveformPointer pWaveform) {
-    m_waveformSummary = pWaveform;
+    if (m_pWaveformSummary == pWaveform) {
```
Comment (daschuer) [inline src/trackinfoobject.cpp]: Move inside the lock.

File: src/trackinfoobject.cpp (hunk 13 above)
Comment (uklotzde) [inline src/trackinfoobject.cpp]: The comparison of the pointers is thread-safe, but the whole set function is 
not atomic. If we acquire the lock after the comparison in very rare cases 
the member pointer might be modified by another thread in the meantime.

**Comment by daschuer** [comment]
I am not sure if we have solved all theoretically race conditions with: 

ConstWaveformPointer m_pWaveform;
ConstWaveformPointer m_pWaveformSummary;
bool m_bClearWaveformRequested;

I have not understand the code entirely, but I can think of a lot of artificial races between these values above. In the current solution you can write and read all these values from all threads at any time. 
What should for instance happen if  m_pWaveform is set while m_bClearWaveformRequested is already set as well? 

What are the cases we have to consider? 

By the way: A bool can be written in on CPU cycle, so there is no need for a lock. However in other places we have uses an QAtomicInt for this. This makes sure that all reviewers knows that the atomic nature is important for the code.

**Comment by uklotzde** [comment]
I wanted to propose a single function for setting all those members at once. 
But I then thought this would be out of scope and cancelled my already 
written comment.

The TIO has lots of issues when setting individual members concurrently from 
different threads. Many interactions with the TIO are not atomic The design 
shows a very naive approach on handling concurrency issues. Improving this 
broken design will take some time ;)

**Comment by daschuer** [comment]
Since most of these issues are probably only theoretically, It would work for me, if we just drop some TODO comments about the issues

**Comment by daschuer** [comment]
I have just notice a real issue: 
* Load a track to Deck A 
* Clear waveforms 
* Load the same track to Deck B 
* Deck A shows mediately the wavefom while Deck B is still analyzing.  

Is the old waveforms showen in Deck A?

**Comment by daschuer** [comment]
I have just retested this branch, unfortunately the issue 
https://github.com/mixxxdj/mixxx/pull/841#issuecomment-170852713
is still there. 
You can watch it the best if you set the cue point to the very end with "load at cue" enabled.

**Comment by ninomp** [comment]
Yes, I am aware of that. It is just that I haven't found time to fix the issue.

**Comment by rryan** [comment]
Howdy @ninomp -- are you still interested in working on this?

**Comment by ninomp** [comment]
Hi @rryan Yes, but it seems that I need more experience to solve all the issues and since I would like to take a different approach with this, I am closing this PR (for now).
